src/core/experimental/memutils.d

```diff
@@ -0,0 +1,685 @@
+/**
+ * Pure D replacement of the C Standard Library basic memory building blocks of string.h
+ * Source: $(DRUNTIMESRC core/experimental/memutils.d)
+ */
+module core.experimental.memutils;
+
+void memmove(T)(T *dst, const T *src)
+{
+    void *d = cast(void*) dst;
+    const(void) *s = cast(void*) src;
+    if ((cast(ulong)d - cast(ulong)s) < T.sizeof)
+    {  // There is overlap with dest being ahead. Use backwards move.
+        Dmemmove_back(d, s, T.sizeof);
+    }
+    else if ((cast(ulong)s - cast(ulong)d) < T.sizeof)
+    {  // There is overlap with src being ahead. Use backwards move.
+        Dmemmove_forw(d, s, T.sizeof);
+    }
+    else
+    {  // There is no overlap, use memcpy.
+        Dmemcpy(dst, src);
+    }
+}
+
+unittest
+{
+    real a = 1.2;
+    real b;
+    memmove(&b, &a);
+    assert(b == 1.2);
+    // Overwrite the type system and create overlap with dst forward.
+    ubyte[8] buf;
+    int *p = cast(int*) buf.ptr;
+    int *q = cast(int*) (buf.ptr + 2);
+    *p = 203847;
+    memmove(q, p);
+    assert(*q == 203847);
+    // Create overlap with src forward.
+    *q = 92239;
+    memmove(p, q);
+    assert(*p == 92239);
+}
+
+/* Dynamic Arrays
+ */
+void memmove(T)(T[] dst, const T[] src)
+{
+    mixin(arrayCode);
+}
+
+/* Static Arrays
+ */
+void memmove(T, size_t len)(ref T[len] dst, ref const T[len] src)
+{
+    mixin(arrayCode);
+}
+
+enum arrayCode = "
+    assert(dst.length == src.length);
+    void *d = cast(void*) dst.ptr;
+    const void *s = cast(const(void)*) src.ptr;
+    size_t n = dst.length * typeof(dst[0]).sizeof;
+    if ((cast(ulong)d - cast(ulong)s) < n)
+    {  // There is overlap with dest being ahead. Use backwards move.
+        Dmemmove_back(d, s, n);
+    }
+    else if ((cast(ulong)s - cast(ulong)d) < n)
+    {  // There is overlap with src being ahead. Use backwards move.
+        Dmemmove_forw(d, s, n);
+    }
+    else
+    {  // There is no overlap, use memcpy.
+        pragma(inline, true);
+        Dmemcpy(d, s, n);
+    }";
+
+
+
+unittest
+{
+    const float[3] a = [1.2, 3.4, 5.8];
+    float[3] b;
+    memmove(b, a);
+    assert(b[0] == 1.2f);
+    assert(b[1] == 3.4f);
+    assert(b[2] == 5.8f);
+}
+
+/* Can we use SIMD?
+ */
+version (D_SIMD)
+{
+    import core.simd: float4;
+    enum useSIMD = true;
+}
+else version (LDC)
+{
+    // LDC always supports SIMD (but doesn't ever set D_SIMD) and
+    // the back-end uses the most appropriate size for every target.
+    import core.simd: float4;
+    enum useSIMD = true;
+}
+else version (GNU)
+{
+    import core.simd: float4;
+    // GNU does not support SIMD by default.
+    version (X86_64)
+    {
+        enum isX86 = true;
+    }
+    else version (X86)
+    {
+        enum isX86 = true;
+    }
+
+    static if (isX86 && __traits(compiles, float4))
+    {
+        enum useSIMD = true;
+    }
+    else
+    {
+        enum useSIMD = false;
+    }
+}
+else
+{
+    enum useSIMD = false;
+}
+
+/* Little SIMD library
+ */
```

Comment:
Maybe move into a separate module?

Reply:
one possibility is to make `core.experimental.mem` a package with modules `cpy`, `move`, and `simd`.

Reply:
Better name it `memory` as there's `core.memory`.

Reply:
Ok, I'm making a new PR. I have seen all the comments, I left them to be resolved in the new PR.